File: day6/day6.cpp
#include <cctype>
#include <fstream>
#include <vector>

#include <fmt/format.h>

using namespace std;

static pair<string,string> error_correct(const vector<string>& messages)
{
	string most, least;
	for (size_t col = 0; col < messages[0].size(); col++)
	{
		vector<pair<int,int>> freq(26);
		for (auto& msg: messages)
		{
			int i = msg[col];
			if (isalpha(i))
			{
				freq[i-'a'].first++;
				freq[i-'a'].second = i;
			}
		}
		sort(freq.begin(), freq.end(), [](const auto &a, const auto& b) {
			return a.first > b.first;
		});
		most.push_back(freq.front().second);
		while (freq.back().first == 0)
		{
			freq.pop_back();
		}
		least.push_back(freq.back().second);
	}
	return make_pair(move(most),move(least));
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<string> messages;
	string message;
	while (getline(input, message))
	{
		messages.emplace_back(message);
	}
	input.close();

	auto [most, least] = error_correct(messages);
	fmt::print("Part1: {}\nPart2: {}\n", most, least);

	return 0;
}

File: day6/day6.py
#!/usr/bin/env python3

def decode(txt):
    rows = txt.split("\n")
    length = len(rows[0])
    most = []
    least = []
    for j in range(length):
        freq = {}
        for row in rows:
            freq[row[j]] = freq.get(row[j], 0) + 1

        lst = sorted(freq.items(), key = lambda x: (-x[1], x[0]))
        most.append(lst[0][0])
        least.append(lst[-1][0])

    return "".join(most), "".join(least)

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    part1, part2 = decode(txt)
    print("Part1:", part1)
    print("Part2:", part2)
